internal/venice-common/src/main/java/com/linkedin/venice/servicediscovery/ServiceDiscoveryAnnouncerRetryTask.java

```diff
@@ -0,0 +1,56 @@
+package com.linkedin.venice.servicediscovery;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class is responsible for retrying the registration of a {@link ServiceDiscoveryAnnouncer} in the case of registration failure.
+ * It is a runnable task that is scheduled to register each service discovery announcer every
+ * {@link ServiceDiscoveryAnnouncerRetryTask#serviceDiscoveryRegistrationRetryMS} milliseconds.
+ */
+public class ServiceDiscoveryAnnouncerRetryTask implements Runnable {
+  private static final Logger LOGGER = LogManager.getLogger(ServiceDiscoveryAnnouncerRetryTask.class);
+  private final BlockingQueue<ServiceDiscoveryAnnouncer> serviceDiscoveryAnnouncerRetryQueue;
+  private final long serviceDiscoveryRegistrationRetryMS;
+
+  public ServiceDiscoveryAnnouncerRetryTask(
+      BlockingQueue<ServiceDiscoveryAnnouncer> serviceDiscoveryAnnouncerRetryQueue,
+      long serviceDiscoveryRegistrationRetryMS) {
+    this.serviceDiscoveryAnnouncerRetryQueue = serviceDiscoveryAnnouncerRetryQueue;
+    this.serviceDiscoveryRegistrationRetryMS = serviceDiscoveryRegistrationRetryMS;
+  }
+
+  /**
+   * If the retry queue is empty, the task will wait until there is an available element in the queue.
+   * Each service discovery announcer will retry registration every {@link ServiceDiscoveryAnnouncerRetryTask#serviceDiscoveryRegistrationRetryMS} milliseconds.
+   */
+  @Override
+  public void run() {
+    while (true) {
```

Comment:
One minor optimization could be that we could change this while true to `!serviceDiscoveryAnnouncerRetryQueue.isEmpty()` - that way, the thread can exit if there are no more failures